File: src/main/java/com/gg/plugins/mongo/view/MongoExplorerPanel.java
/*
 * Copyright (c) 2022. Geetesh Gupta
 */

package com.gg.plugins.mongo.view;

import com.gg.plugins.mongo.action.explorer.*;
import com.gg.plugins.mongo.config.ConfigurationException;
import com.gg.plugins.mongo.config.MongoConfiguration;
import com.gg.plugins.mongo.config.ServerConfiguration;
import com.gg.plugins.mongo.editor.MongoFileSystem;
import com.gg.plugins.mongo.editor.MongoObjectFile;
import com.gg.plugins.mongo.model.*;
import com.gg.plugins.mongo.model.navigation.Navigation;
import com.gg.plugins.mongo.service.MongoService;
import com.gg.plugins.mongo.service.Notifier;
import com.gg.plugins.mongo.utils.GuiUtils;
import com.intellij.ide.CommonActionsManager;
import com.intellij.ide.TreeExpander;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.actionSystem.ActionManager;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.DefaultActionGroup;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.MessageType;
import com.intellij.openapi.ui.Splitter;
import com.intellij.openapi.ui.popup.Balloon;
import com.intellij.openapi.util.Disposer;
import com.intellij.ui.DoubleClickListener;
import com.intellij.ui.PopupHandler;
import com.intellij.ui.TreeSpeedSearch;
import com.intellij.ui.components.JBScrollPane;
import com.intellij.ui.treeStructure.Tree;
import com.intellij.util.ui.UIUtil;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import javax.swing.tree.TreePath;
import javax.swing.tree.TreeSelectionModel;
import java.awt.*;
import java.awt.event.MouseEvent;
import java.net.URL;
import java.util.Set;

import static com.gg.plugins.mongo.utils.GuiUtils.showNotification;

public class MongoExplorerPanel extends JPanel implements Disposable {

	private static final URL pluginSettingsUrl = GuiUtils.class.getResource("/general/add.png");

	private final Project project;

	private final MongoService mongoService;

	private final Notifier notifier;

	private final JPanel treePanel;

	private final Tree mongoTree;
	//
	//	private final MongoTreeBuilder mongoTreeModel;

	private final MongoTreeModel mongoTreeModel;

	private JPanel rootPanel;

	private JPanel toolBarPanel;

	private JPanel containerPanel;

	public MongoExplorerPanel(Project project) {
		this.project = project;
		this.mongoService = MongoService.getInstance(project);
		this.notifier = Notifier.getInstance(project);

		treePanel = new JPanel(new BorderLayout());
		treePanel.setLayout(new BorderLayout());

		mongoTree = createTree();
		mongoTreeModel = new MongoTreeModel();
		mongoTree.setModel(mongoTreeModel);

		setLayout(new BorderLayout());
		treePanel.add(new JBScrollPane(mongoTree), BorderLayout.CENTER);

		Splitter splitter = new Splitter(true, 0.6f);
		splitter.setFirstComponent(treePanel);

		containerPanel.add(splitter, BorderLayout.CENTER);

		add(rootPanel, BorderLayout.CENTER);

		toolBarPanel.setLayout(new BorderLayout());

		init();
	}

	private Tree createTree() {

		Tree tree = new Tree() {

			private final JLabel myLabel = new JLabel(String.format(
					"<html><center>No Mongo server available<br><br>You may use <img src=\"%s\"> to add " +
					"configuration</center></html>",
					pluginSettingsUrl));

			@Override
			protected void paintComponent(Graphics g) {
				super.paintComponent(g);
				if (!getServerConfigurations().isEmpty())
					return;

				myLabel.setFont(getFont());
				myLabel.setBackground(getBackground());
				myLabel.setForeground(getForeground());
				Rectangle bounds = getBounds();
				Dimension size = myLabel.getPreferredSize();
				myLabel.setBounds(0, 0, size.width, size.height);

				int x = (bounds.width - size.width) / 2;
				Graphics g2 = g.create(bounds.x + x, bounds.y + 20, bounds.width, bounds.height);
				try {
					myLabel.paint(g2);
				} finally {
					g2.dispose();
				}
			}
		};

		tree.getEmptyText().clear();
		tree.getSelectionModel().setSelectionMode(TreeSelectionModel.SINGLE_TREE_SELECTION);
		tree.setName("mongoTree");
		tree.setRootVisible(false);

		new TreeSpeedSearch(tree, treePath -> {
			final Object node = treePath.getLastPathComponent();
			if (node instanceof MongoDatabase) {
				return ((MongoDatabase) node).getName();
			}
			if (node instanceof MongoCollection) {
				return ((MongoCollection) node).getName();
			}
			return "<empty>";
		});

		return tree;
	}

	public void init() {
		loadAllServerConfigurations();
		installActions();
		doubleClickEventHandler();
	}

	private Set<ServerConfiguration> getServerConfigurations() {
		return MongoConfiguration.getInstance(project).getServerConfigurations();
	}

	private void loadAllServerConfigurations() {
		this.mongoService.cleanUpServers();

		Set<ServerConfiguration> serverConfigurations = getServerConfigurations();
		for (ServerConfiguration serverConfiguration : serverConfigurations) {
			addConfiguration(serverConfiguration);
		}
	}

	private void installActions() {
		Actions actions = new Actions();
		Disposer.register(this, () -> {
			actions.collapseAllAction.unregisterCustomShortcutSet(rootPanel);
			actions.expandAllAction.unregisterCustomShortcutSet(rootPanel);
		});

		DefaultActionGroup actionGroup = new DefaultActionGroup("MongoExplorerGroup", false);

		if (ApplicationManager.getApplication() != null) {
			actionGroup.add(actions.addServerAction);
			actionGroup.add(actions.duplicateServerAction);
			actionGroup.addSeparator();
			actionGroup.add(actions.refreshServerAction);
			actionGroup.add(actions.mongoConsoleAction);
			actionGroup.addSeparator();
			actionGroup.add(actions.expandAllAction);
			actionGroup.add(actions.collapseAllAction);
			actionGroup.addSeparator();
			actionGroup.add(actions.openPluginSettingsAction);
		}
		GuiUtils.installActionGroupInToolBar(actionGroup,
				toolBarPanel,
				ActionManager.getInstance(),
				"MongoExplorerActions",
				true);

		DefaultActionGroup actionPopupGroup = new DefaultActionGroup("MongoExplorerPopupGroup", true);
		if (ApplicationManager.getApplication() != null) {
			actionPopupGroup.add(actions.refreshServerAction);
			actionPopupGroup.add(actions.editServerAction);
			actionPopupGroup.add(actions.duplicateServerAction);
			actionPopupGroup.add(actions.deleteAction);
			actionPopupGroup.addSeparator();
			actionPopupGroup.add(actions.viewCollectionValuesAction);
		}
		PopupHandler.installPopupMenu(mongoTree, actionPopupGroup, "POPUP");
	}

	private void doubleClickEventHandler() {
		new DoubleClickListener() {
			@Override
			protected boolean onDoubleClick(@NotNull MouseEvent event) {
				if (!(event.getSource() instanceof JTree)) {
					return false;
				}

				Object node = getSelectedNode();

				if (node instanceof MongoServer && ((MongoServer) node).getDatabases().isEmpty()) {
					openServer((MongoServer) node);
				} else if (node instanceof MongoDatabase && ((MongoDatabase) node).getCollections().isEmpty()) {
					loadDatabase((MongoDatabase) node);
				} else if (node instanceof MongoCollection) {
					loadSelectedCollectionValues((MongoCollection) node);
				}
				return false;
			}
		}.installOn(mongoTree);
	}

	public void addConfiguration(ServerConfiguration serverConfiguration) {
		MongoServer mongoServer = mongoTreeModel.addConfiguration(serverConfiguration);
		mongoService.registerServer(mongoServer);
	}

	public Object getSelectedNode() {
		return mongoTree.getLastSelectedPathComponent();
	}

	public void openServer(final MongoServer mongoServer) {
		ProgressManager.getInstance().run(new Task.Backgroundable(project, "Connecting to " + mongoServer.getLabel()) {

			@Override
			public void run(@NotNull ProgressIndicator indicator) {
				mongoTree.setPaintBusy(true);
				ApplicationManager.getApplication().invokeLater(() -> {
					try {
						Set<MongoDatabase> mongoDatabases = mongoService.loadDatabases(mongoServer,
								mongoTreeModel.getServerConfiguration(mongoServer));
						if (mongoDatabases.isEmpty()) {
							return;
						}
						mongoServer.setDatabases(mongoDatabases);
						mongoTreeModel.fireTreeStructureChanged(mongoServer);
					} catch (ConfigurationException confEx) {
						mongoServer.setStatus(MongoServer.Status.ERROR);

						String errorMessage = String.format("Error when connecting to %s", mongoServer.getLabel());
						notifier.notifyError(errorMessage + ": " + confEx.getMessage());
						UIUtil.invokeLaterIfNeeded(() -> showNotification(treePanel,
								MessageType.ERROR,
								errorMessage,
								Balloon.Position.atLeft));
					} catch (Exception e) {
						throw new RuntimeException(e);
					} finally {
						mongoTree.setPaintBusy(false);
					}
				});
			}
		});
	}

	public void loadDatabase(MongoDatabase mongoDatabase) {
		ServerConfiguration configuration = mongoTreeModel.getServerConfiguration(mongoDatabase);
		Set<MongoCollection> collections = mongoService.loadCollections(mongoDatabase, configuration);
		mongoDatabase.setCollections(collections);
		mongoTreeModel.fireTreeStructureChanged(mongoDatabase);
	}

	public void loadSelectedCollectionValues(MongoCollection mongoCollection) {
		MongoServer parentServer = mongoCollection.getParentDatabase().getParentServer();
		ServerConfiguration configuration = parentServer.getConfiguration();

		Navigation navigation = new Navigation();
		MongoQueryOptions queryOptions = new MongoQueryOptions();
		queryOptions.setResultLimit(configuration.getDefaultRowLimit());
		navigation.addNewWayPoint(mongoCollection, queryOptions);

		MongoFileSystem.getInstance().openEditor(new MongoObjectFile(project, configuration, navigation));
	}

	public void removeNode(Object node) {
		if (node instanceof MongoServer)
			mongoTreeModel.removeConfiguration((MongoServer) node);
		else if (node instanceof MongoDatabase) {
			mongoService.removeDatabase(getServerConfiguration(node), (MongoDatabase) node);
			openServer(((MongoDatabase) node).getParentServer());
		} else if (node instanceof MongoCollection) {
			mongoService.removeCollection(getServerConfiguration(node), (MongoCollection) node);
			loadDatabase(((MongoCollection) node).getParentDatabase());
		}
		mongoTreeModel.fireTreeStructureChanged(mongoTreeModel.getParent(node));
	}

	public ServerConfiguration getServerConfiguration(Object treeNode) {
		return mongoTreeModel.getServerConfiguration(treeNode);
	}

	@Override
	public void dispose() {

	}

	private void expandAll() {
		if (mongoTreeModel.getChildCount(mongoTreeModel.getRoot()) > 0)
			mongoTreeModel.getChildren(mongoTreeModel.getRoot())
			              .forEach(c -> mongoTree.expandPath(new TreePath(mongoTreeModel.getPathToRoot(c))));
	}

	public String getTypeOfNode(Object node) {
		return mongoTreeModel.getTypeOfNode(node);
	}

	private void collapseAll() {
		if (mongoTreeModel.getChildCount(mongoTreeModel.getRoot()) > 0)
			mongoTreeModel.getChildren(mongoTreeModel.getRoot())
			              .forEach(c -> mongoTree.collapsePath(new TreePath(mongoTreeModel.getPathToRoot(c))));
	}

	public JPanel getContent() {
		return rootPanel;
	}

	private class Actions {
		public final RefreshServerAction refreshServerAction = new RefreshServerAction(MongoExplorerPanel.this);

		public final AddServerAction addServerAction = new AddServerAction(MongoExplorerPanel.this);

		public final DuplicateServerAction duplicateServerAction = new DuplicateServerAction(MongoExplorerPanel.this);

		public final EditServerAction editServerAction = new EditServerAction(MongoExplorerPanel.this);

		public final DeleteAction deleteAction = new DeleteAction(MongoExplorerPanel.this);

		public final ViewCollectionValuesAction viewCollectionValuesAction =
				new ViewCollectionValuesAction(MongoExplorerPanel.this);

		public final OpenPluginSettingsAction openPluginSettingsAction = new OpenPluginSettingsAction();

		public final MongoConsoleAction mongoConsoleAction = new MongoConsoleAction(MongoExplorerPanel.this);

		final TreeExpander treeExpander = new TreeExpander() {
			@Override
			public void expandAll() {
				MongoExplorerPanel.this.expandAll();
			}

			@Override
			public boolean canExpand() {
				return !getServerConfigurations().isEmpty();
			}

			@Override
			public void collapseAll() {
				MongoExplorerPanel.this.collapseAll();
			}

			@Override
			public boolean canCollapse() {
				return !getServerConfigurations().isEmpty();
			}
		};

		final CommonActionsManager actionsManager = CommonActionsManager.getInstance();

		public final AnAction expandAllAction = actionsManager.createExpandAllAction(treeExpander, rootPanel);

		public final AnAction collapseAllAction = actionsManager.createCollapseAllAction(treeExpander, rootPanel);
	}

}

File: src/main/kotlin/com/gg/plugins/mongo/view/MongoExplorerPanel.kt
/*
 * Copyright (c) 2018 David Boissier.
 * Modifications Copyright (c) 2022 Geetesh Gupta.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.gg.plugins.mongo.view

import com.gg.plugins.mongo.action.explorer.AddServerAction
import com.gg.plugins.mongo.action.explorer.DeleteAction
import com.gg.plugins.mongo.action.explorer.DuplicateServerAction
import com.gg.plugins.mongo.action.explorer.EditServerAction
import com.gg.plugins.mongo.action.explorer.MongoConsoleAction
import com.gg.plugins.mongo.action.explorer.OpenPluginSettingsAction
import com.gg.plugins.mongo.action.explorer.RefreshServerAction
import com.gg.plugins.mongo.action.explorer.ViewCollectionValuesAction
import com.gg.plugins.mongo.config.ConfigurationException
import com.gg.plugins.mongo.config.MongoConfiguration
import com.gg.plugins.mongo.config.ServerConfiguration
import com.gg.plugins.mongo.editor.MongoFileSystem
import com.gg.plugins.mongo.editor.MongoObjectFile
import com.gg.plugins.mongo.model.MongoCollection
import com.gg.plugins.mongo.model.MongoConstants
import com.gg.plugins.mongo.model.MongoDatabase
import com.gg.plugins.mongo.model.MongoQueryOptions
import com.gg.plugins.mongo.model.MongoServer
import com.gg.plugins.mongo.model.MongoTreeModel
import com.gg.plugins.mongo.model.navigation.Navigation
import com.gg.plugins.mongo.service.MongoService
import com.gg.plugins.mongo.service.Notifier
import com.intellij.ide.CommonActionsManager
import com.intellij.ide.TreeExpander
import com.intellij.openapi.Disposable
import com.intellij.openapi.actionSystem.ActionManager
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.DefaultActionGroup
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.progress.ProgressIndicator
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.progress.Task
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.MessageType
import com.intellij.openapi.ui.Splitter
import com.intellij.openapi.ui.popup.Balloon
import com.intellij.openapi.util.Disposer
import com.intellij.ui.ColoredTreeCellRenderer
import com.intellij.ui.DoubleClickListener
import com.intellij.ui.GuiUtils
import com.intellij.ui.PopupHandler
import com.intellij.ui.TreeSpeedSearch
import com.intellij.ui.components.JBScrollPane
import com.intellij.ui.treeStructure.Tree
import com.intellij.util.containers.Convertor
import com.intellij.util.ui.UIUtil
import java.awt.BorderLayout
import java.awt.Graphics
import java.awt.event.MouseEvent
import java.net.URL
import java.util.TreeSet
import java.util.function.Consumer
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JTree
import javax.swing.plaf.basic.BasicTreeUI
import javax.swing.tree.TreePath
import javax.swing.tree.TreeSelectionModel

class MongoExplorerPanel(private val project: Project) : JPanel(), Disposable {
    private val mongoService: MongoService = MongoService.getInstance(project)
    private val notifier: Notifier = Notifier.getInstance(project)
    private val treePanel: JPanel = JPanel(BorderLayout())
    private val mongoTreeModel: MongoTreeModel
    private val mongoTree: Tree
    lateinit var rootPanel: JPanel
    lateinit var toolBarPanel: JPanel
    lateinit var containerPanel: JPanel

    init {
        treePanel.layout = BorderLayout()
        mongoTree = createTree()
        mongoTreeModel = MongoTreeModel()
        mongoTree.model = mongoTreeModel
        layout = BorderLayout()
        treePanel.add(JBScrollPane(mongoTree), BorderLayout.CENTER)
        val splitter = Splitter(true, 0.6f)
        splitter.firstComponent = treePanel
        containerPanel.add(splitter, BorderLayout.CENTER)
        add(rootPanel, BorderLayout.CENTER)
        toolBarPanel.layout = BorderLayout()

        loadAllServerConfigurations()
        installActions()
        doubleClickEventHandler()
    }

    private fun createTree(): Tree {
        val tree: Tree = object : Tree() {
            private val myLabel = JLabel(
                String.format(
                    "<html><center>No Mongo server available<br><br>You may use <img src=\"%s\"> to add " +
                            "configuration</center></html>",
                    pluginSettingsUrl
                )
            )

            override fun paintComponent(g: Graphics) {
                super.paintComponent(g)
                if (serverConfigurations.isNotEmpty()) return
                myLabel.font = font
                myLabel.background = background
                myLabel.foreground = foreground
                val bounds = bounds
                val size = myLabel.preferredSize
                myLabel.setBounds(0, 0, size.width, size.height)
                val x = (bounds.width - size.width) / 2
                val g2 = g.create(bounds.x + x, bounds.y + 20, bounds.width, bounds.height)
                try {
                    myLabel.paint(g2)
                } finally {
                    g2.dispose()
                }
            }
        }
        tree.setUI(object : BasicTreeUI() {
            override fun shouldPaintExpandControl(
                path: TreePath,
                row: Int,
                isExpanded: Boolean,
                hasBeenExpanded: Boolean,
                isLeaf: Boolean
            ): Boolean {
                val node = path.lastPathComponent
                return mongoTreeModel.getChildren(node).isNotEmpty()
            }
        })
        tree.emptyText.clear()
        tree.selectionModel.selectionMode = TreeSelectionModel.SINGLE_TREE_SELECTION
        tree.name = "mongoTree"
        tree.isRootVisible = false
        tree.cellRenderer = object : ColoredTreeCellRenderer() {
            override fun customizeCellRenderer(
                tree: JTree,
                value: Any,
                selected: Boolean,
                expanded: Boolean,
                leaf: Boolean,
                row: Int,
                hasFocus: Boolean
            ) {
                when (value) {
                    is MongoServer -> {
                        icon = MongoConstants.MONGO_SERVER
                        toolTipText = "Server: " + value.label
                        append(value.label)
                    }

                    is MongoDatabase -> {
                        icon = MongoConstants.MONGO_DATABASE
                        toolTipText = "Database: " + value.name
                        append(value.name)
                    }

                    is MongoCollection -> {
                        icon = MongoConstants.MONGO_COLLECTION
                        toolTipText = "Collection: " + value.name
                        append(value.name)
                    }
                }
            }
        }
        TreeSpeedSearch(tree, Convertor { treePath: TreePath ->
            val node = treePath.lastPathComponent
            if (node is MongoDatabase) {
                return@Convertor node.name
            }
            if (node is MongoCollection) {
                return@Convertor node.name
            }
            "<empty>"
        })
        return tree
    }

    private val serverConfigurations: Set<ServerConfiguration>
        get() = MongoConfiguration.getInstance(project).serverConfigurations

    private fun loadAllServerConfigurations() {
        mongoService.cleanUpServers()
        for (serverConfiguration in serverConfigurations) {
            addConfiguration(serverConfiguration)
        }
    }

    private fun installActions() {
        val actions = Actions()
        Disposer.register(this) {
            actions.collapseAllAction.unregisterCustomShortcutSet(rootPanel)
            actions.expandAllAction.unregisterCustomShortcutSet(rootPanel)
        }
        val actionGroup = DefaultActionGroup("MongoExplorerGroup", false)
        if (ApplicationManager.getApplication() != null) {
            actionGroup.add(actions.addServerAction)
            actionGroup.add(actions.duplicateServerAction)
            actionGroup.addSeparator()
            actionGroup.add(actions.refreshServerAction)
            actionGroup.add(actions.mongoConsoleAction)
            actionGroup.addSeparator()
            actionGroup.add(actions.expandAllAction)
            actionGroup.add(actions.collapseAllAction)
            actionGroup.addSeparator()
            actionGroup.add(actions.openPluginSettingsAction)
        }
        com.gg.plugins.mongo.utils.GuiUtils.installActionGroupInToolBar(
            actionGroup,
            toolBarPanel,
            ActionManager.getInstance(),
            "MongoExplorerActions",
            true
        )
        val actionPopupGroup = DefaultActionGroup("MongoExplorerPopupGroup", true)
        if (ApplicationManager.getApplication() != null) {
            actionPopupGroup.add(actions.refreshServerAction)
            actionPopupGroup.add(actions.editServerAction)
            actionPopupGroup.add(actions.duplicateServerAction)
            actionPopupGroup.add(actions.deleteAction)
            actionPopupGroup.addSeparator()
            actionPopupGroup.add(actions.viewCollectionValuesAction)
        }
        PopupHandler.installPopupMenu(mongoTree, actionPopupGroup, "POPUP")
    }

    private fun doubleClickEventHandler() {
        object : DoubleClickListener() {
            override fun onDoubleClick(event: MouseEvent): Boolean {
                if (event.source !is JTree) {
                    return false
                }
                val node: Any? = selectedNode
                if (node is MongoServer && node.databases.isEmpty()) {
                    openServer(node)
                } else if (node is MongoDatabase && node.collections.isEmpty()) {
                    loadDatabase(node)
                } else if (node is MongoCollection) {
                    loadSelectedCollectionValues(node)
                }
                return false
            }
        }.installOn(mongoTree)
    }

    fun addConfiguration(serverConfiguration: ServerConfiguration) {
        val mongoServer = mongoTreeModel.addConfiguration(serverConfiguration)
        mongoService.registerServer(mongoServer)
    }

    val selectedNode: Any?
        get() = mongoTree.lastSelectedPathComponent

    fun openServer(mongoServer: MongoServer) {
        ProgressManager.getInstance().run(object : Task.Backgroundable(
            project, "Connecting to " + mongoServer.label
        ) {
            override fun run(indicator: ProgressIndicator) {
                mongoTree.setPaintBusy(true)
                ApplicationManager.getApplication().invokeLater {
                    try {
                        val mongoDatabases = mongoService.loadDatabases(
                            mongoServer,
                            mongoServer.configuration
                        )
                        if (mongoDatabases.isEmpty()) {
                            return@invokeLater
                        }
                        mongoServer.databases = mongoDatabases
                        mongoTreeModel.fireTreeStructureChanged(mongoServer)
                    } catch (confEx: ConfigurationException) {
                        mongoServer.status = MongoServer.Status.ERROR
                        val errorMessage = String.format("Error when connecting to %s", mongoServer.label)
                        notifier.notifyError(errorMessage + ": " + confEx.message)
                        UIUtil.invokeLaterIfNeeded {
                            com.gg.plugins.mongo.utils.GuiUtils.showNotification(
                                treePanel,
                                MessageType.ERROR,
                                errorMessage,
                                Balloon.Position.atLeft
                            )
                        }
                    } catch (e: Exception) {
                        throw RuntimeException(e)
                    } finally {
                        mongoTree.setPaintBusy(false)
                    }
                }
            }
        })
    }

    fun loadDatabase(mongoDatabase: MongoDatabase) {
        val configuration = mongoDatabase.parentServer.configuration
        val collections = mongoService.loadCollections(mongoDatabase, configuration)
        mongoDatabase.collections = collections as TreeSet<MongoCollection>
        mongoTreeModel.fireTreeStructureChanged(mongoDatabase)
    }

    fun loadSelectedCollectionValues(mongoCollection: MongoCollection) {
        val parentServer = mongoCollection.parentDatabase.parentServer
        val configuration = parentServer.configuration
        val navigation = Navigation()
        val queryOptions = MongoQueryOptions()
        queryOptions.resultLimit = configuration.defaultRowLimit
        navigation.addNewWayPoint(mongoCollection, queryOptions)
        MongoFileSystem.getInstance().openEditor(MongoObjectFile(project, configuration, navigation))
    }

    fun removeNode(node: Any) {
        when (node) {
            is MongoServer -> mongoTreeModel.removeConfiguration(node)
            is MongoDatabase -> {
                mongoService.removeDatabase(node.parentServer.configuration, node)
                openServer(node.parentServer)
            }

            is MongoCollection -> {
                mongoService.removeCollection(node.parentDatabase.parentServer.configuration, node)
                loadDatabase(node.parentDatabase)
            }
        }
        mongoTreeModel.fireTreeStructureChanged(mongoTreeModel.getParent(node))
    }

    fun getServerConfiguration(treeNode: Any?): ServerConfiguration? {
        return mongoTreeModel.getServerConfiguration(treeNode)
    }

    override fun dispose() {}
    private fun expandAll() {
        if (mongoTreeModel.getChildCount(mongoTreeModel.root) > 0) mongoTreeModel.getChildren(mongoTreeModel.root)
            .forEach(Consumer { c: Any? -> mongoTree.expandPath(TreePath(mongoTreeModel.getPathToRoot(c))) })
    }

    fun getTypeOfNode(node: Any?): String {
        return mongoTreeModel.getTypeOfNode(node)
    }

    private fun collapseAll() {
        if (mongoTreeModel.getChildCount(mongoTreeModel.root) > 0) mongoTreeModel.getChildren(mongoTreeModel.root)
            .forEach(Consumer { c: Any? -> mongoTree.collapsePath(TreePath(mongoTreeModel.getPathToRoot(c))) })
    }

    private inner class Actions {
        val refreshServerAction = RefreshServerAction(this@MongoExplorerPanel)
        val addServerAction = AddServerAction(this@MongoExplorerPanel)
        val duplicateServerAction = DuplicateServerAction(this@MongoExplorerPanel)
        val editServerAction = EditServerAction(this@MongoExplorerPanel)
        val deleteAction = DeleteAction(this@MongoExplorerPanel)
        val viewCollectionValuesAction = ViewCollectionValuesAction(this@MongoExplorerPanel)
        val openPluginSettingsAction = OpenPluginSettingsAction()
        val mongoConsoleAction = MongoConsoleAction(this@MongoExplorerPanel)
        val treeExpander: TreeExpander = object : TreeExpander {
            override fun expandAll() {
                this@MongoExplorerPanel.expandAll()
            }

            override fun canExpand(): Boolean {
                return serverConfigurations.isNotEmpty()
            }

            override fun collapseAll() {
                this@MongoExplorerPanel.collapseAll()
            }

            override fun canCollapse(): Boolean {
                return serverConfigurations.isNotEmpty()
            }
        }
        val actionsManager: CommonActionsManager = CommonActionsManager.getInstance()
        val expandAllAction: AnAction = actionsManager.createExpandAllAction(treeExpander, rootPanel)
        val collapseAllAction: AnAction = actionsManager.createCollapseAllAction(treeExpander, rootPanel)
    }

    companion object {
        private val pluginSettingsUrl: URL = GuiUtils::class.java.getResource("/general/add.png")!!
    }
}